File: benches/fixtures/sample.py
"""Sample Python module for benchmarking Taoki's index_source."""

from __future__ import annotations

import os
import sys
import json
import logging
from typing import Any, Dict, List, Optional, Protocol, TypeVar
from dataclasses import dataclass, field
from pathlib import Path
from flask import Flask, Blueprint, request, jsonify

logger = logging.getLogger(__name__)

T = TypeVar("T")

MAX_ITEMS = 1000
DEFAULT_PAGE_SIZE = 20

app = Flask(__name__)
api_bp = Blueprint("api", __name__, url_prefix="/api/v1")


class ServiceError(Exception):
    """Base exception for service errors."""

    def __init__(self, message: str, code: int = 500) -> None:
        super().__init__(message)
        self.code = code


class NotFoundError(ServiceError):
    """Raised when a resource is not found."""

    def __init__(self, resource: str) -> None:
        super().__init__(f"{resource} not found", code=404)


class Indexable(Protocol):
    """Protocol for indexable items."""

    @property
    def id(self) -> str: ...

    @property
    def kind(self) -> str: ...

    def score(self) -> float: ...


@dataclass
class ClientConfig:
    """Configuration for the HTTP client."""

    base_url: str
    timeout_ms: int = 5000
    max_retries: int = 3
    headers: Dict[str, str] = field(default_factory=dict)
    user_agent: str = "taoki/0.1"
    follow_redirects: bool = True
    verify_ssl: bool = True
    proxy: Optional[str] = None


@dataclass
class User:
    id: str
    name: str
    email: str
    role: str = "viewer"

    def to_dict(self) -> Dict[str, Any]:
        return {"id": self.id, "name": self.name, "email": self.email, "role": self.role}

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "User":
        return cls(
            id=data["id"],
            name=data["name"],
            email=data["email"],
            role=data.get("role", "viewer"),
        )


def fetch_user(user_id: str) -> User:
    """Fetch a user by ID."""
    if not user_id:
        raise ServiceError("user_id cannot be empty", code=400)
    return User(id=user_id, name="Alice", email="alice@example.com")


def paginate(items: List[T], page: int = 1, per_page: int = DEFAULT_PAGE_SIZE) -> Dict[str, Any]:
    """Return a paginated slice of items."""
    total = len(items)
    start = (page - 1) * per_page
    end = min(start + per_page, total)
    return {
        "items": items[start:end],
        "total": total,
        "page": page,
        "per_page": per_page,
    }


def parse_header(raw: str) -> Optional[tuple[str, str]]:
    if ":" not in raw:
        return None
    key, _, value = raw.partition(":")
    return key.strip(), value.strip()


def _internal_hash(data: bytes) -> int:
    h = 0xCBF29CE484222325
    for b in data:
        h ^= b
        h = (h * 0x100000001B3) & 0xFFFFFFFFFFFFFFFF
    return h


@api_bp.route("/users/<user_id>", methods=["GET"])
def get_user(user_id: str):
    try:
        user = fetch_user(user_id)
        return jsonify(user.to_dict())
    except NotFoundError as e:
        return jsonify({"error": str(e)}), 404


@api_bp.route("/users", methods=["POST"])
def create_user():
    data = request.get_json()
    user = User.from_dict(data)
    return jsonify(user.to_dict()), 201


app.register_blueprint(api_bp)


def test_fetch_user():
    user = fetch_user("u1")
    assert user.id == "u1"


def test_paginate():
    items = list(range(10))
    result = paginate(items, page=2, per_page=3)
    assert result["items"] == [3, 4, 5]


if __name__ == "__main__":
    app.run(host="0.0.0.0", port=int(os.environ.get("PORT", 8080)))

File: benches/fixtures/sample.rs
use std::collections::HashMap;
use std::fmt;
use std::io::{self, Read, Write};

pub const MAX_RETRIES: usize = 3;
pub const DEFAULT_TIMEOUT_MS: u64 = 5000;

/// Error types for the service layer.
#[derive(Debug, thiserror::Error)]
pub enum ServiceError {
    #[error("not found: {0}")]
    NotFound(String),
    #[error("unauthorized")]
    Unauthorized,
    #[error("io error: {0}")]
    Io(#[from] io::Error),
    #[error("invalid input: {0}")]
    InvalidInput(String),
}

/// Configuration for the HTTP client.
#[derive(Debug, Clone)]
pub struct ClientConfig {
    pub base_url: String,
    pub timeout_ms: u64,
    pub max_retries: usize,
    pub headers: HashMap<String, String>,
    pub user_agent: String,
    pub follow_redirects: bool,
    pub verify_ssl: bool,
    pub proxy: Option<String>,
}

impl Default for ClientConfig {
    fn default() -> Self {
        Self {
            base_url: String::new(),
            timeout_ms: DEFAULT_TIMEOUT_MS,
            max_retries: MAX_RETRIES,
            headers: HashMap::new(),
            user_agent: "taoki/0.1".to_string(),
            follow_redirects: true,
            verify_ssl: true,
            proxy: None,
        }
    }
}

/// A paginated response wrapper.
#[derive(Debug, Clone)]
pub struct Page<T> {
    pub items: Vec<T>,
    pub total: usize,
    pub page: usize,
    pub per_page: usize,
}

impl<T: fmt::Display> fmt::Display for Page<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Page {}/{}", self.page, self.total.div_ceil(self.per_page))
    }
}

/// Trait for items that can be indexed.
pub trait Indexable {
    fn id(&self) -> &str;
    fn kind(&self) -> &str;
    fn score(&self) -> f64;
}

/// Trait for serializable data.
pub trait Serialize {
    fn to_json(&self) -> String;
    fn to_bytes(&self) -> Vec<u8>;
}

/// A user record.
#[derive(Debug, Clone)]
pub struct User {
    pub id: String,
    pub name: String,
    pub email: String,
    pub role: Role,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Role {
    Admin,
    Editor,
    Viewer,
}

impl Indexable for User {
    fn id(&self) -> &str { &self.id }
    fn kind(&self) -> &str { "user" }
    fn score(&self) -> f64 { 1.0 }
}

pub fn fetch_user(id: &str) -> Result<User, ServiceError> {
    if id.is_empty() {
        return Err(ServiceError::InvalidInput("id cannot be empty".to_string()));
    }
    Ok(User {
        id: id.to_string(),
        name: "Alice".to_string(),
        email: "alice@example.com".to_string(),
        role: Role::Viewer,
    })
}

pub fn paginate<T>(items: Vec<T>, page: usize, per_page: usize) -> Page<T> {
    let total = items.len();
    let start = page.saturating_sub(1) * per_page;
    let end = (start + per_page).min(total);
    Page { items: items[start..end].to_vec(), total, page, per_page }
}

pub(crate) fn parse_header(raw: &str) -> Option<(String, String)> {
    raw.split_once(':').map(|(k, v)| (k.trim().to_string(), v.trim().to_string()))
}

fn internal_hash(data: &[u8]) -> u64 {
    let mut h: u64 = 0xcbf29ce484222325;
    for &b in data {
        h ^= b as u64;
        h = h.wrapping_mul(0x100000001b3);
    }
    h
}

macro_rules! impl_display {
    ($t:ty, $fmt:expr) => {
        impl fmt::Display for $t {
            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
                write!(f, $fmt, self)
            }
        }
    };
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fetch_user_ok() {
        let u = fetch_user("u1").unwrap();
        assert_eq!(u.id, "u1");
    }

    #[test]
    fn fetch_user_empty_id() {
        assert!(fetch_user("").is_err());
    }

    #[test]
    fn paginate_basic() {
        let items: Vec<i32> = (0..10).collect();
        let page = paginate(items, 2, 3);
        assert_eq!(page.items, vec![3, 4, 5]);
    }
}
